### ISSUE TITLE
management.metrics.export.tags-as-prefix property seems not to be honored in Spring Boot 2.0.0.RELEASE

### ISSUE THREAD

**Issue by jceloi** [spring-projects/spring-boot#12557]
Greetings,

I'm currently migrating a 1.5.X Spring Boot project to 2.0.0.RELEASE and did struggle a bit on the metrics to micrometer side.

It seems to me that https://github.com/spring-projects/spring-boot/issues/12262 is still relevant.
I have to manually override the HierarchicalNameMapper to a GraphiteHierarchicalNameMapper so as to use the tags as prefix feature (tested and working).

```
    @Bean
    public HierarchicalNameMapper hierarchicalNameMapper() {
        return new GraphiteHierarchicalNameMapper("myTag");
    }
```

The HierarchicalNameMapper.DEFAULT.toHierarchicalName() does not use this property at all and is in my case used by default.

**It is as if there was some missing autoconfiguration somewhere that would not use the correct name mapper in a Graphite context.**

As a result the management.metrics.export.graphite.tags-as-prefix property seems totally useless by default in my context, which is a Graphite context. I can see that it is correctly valued in the MeterRegistry's config and such, but as the DEFAULT Name Mapper is used, it's to no avail.

**Comment by jceloi** [comment]
Looking at GraphiteMetricsExportAutoConfiguration and the matched /condition in my app 

```
	@Bean
	@ConditionalOnMissingBean
	public HierarchicalNameMapper hierarchicalNameMapper() {
		return HierarchicalNameMapper.DEFAULT;
	}
```
I think that the default Name Mapper should be logically a GraphiteHierarchicalNameMapper to use the specificity of that property.

**Comment by wilkinsona** [comment]
> I think that the default Name Mapper should be logically a GraphiteHierarchicalNameMapper to use the specificity of that property.

That certainly sounds logical, although Micrometer's [doing the same thing as Boot](https://github.com/micrometer-metrics/micrometer/blob/cce4e582133617180dc0fa8724df2e9e0b4177ca/micrometer-spring-legacy/src/main/java/io/micrometer/spring/autoconfigure/export/graphite/GraphiteMetricsExportAutoConfiguration.java#L68) in its Spring Boot 1.5 support.

@jkschneider Is this a bug in both codebases?

**Comment by ielfatmi** [comment]
I think that GraphiteMeterRegistry bean instantiation in GraphiteMetricsExportAutoConfiguration should be reviewed to get something like this :

@Bean
@ConditionalOnMissingBean
public GraphiteMeterRegistry graphiteMeterRegistry(GraphiteConfig graphiteConfig, Clock clock) {
  return new GraphiteMeterRegistry(graphiteConfig, clock);
}

Since GraphiteMeterRegistry(GraphiteConfig config, Clock clock) constructor is using GraphiteHierarchicalNameMapper as HierarchicalNameMapper

**Comment by jkschneider** [comment]
It appears to me you are simply missing a word in the property?

`management.metrics.export.graphite.tags-as-prefix` instead of
`management.metrics.export.tags-as-prefix`

**Comment by jceloi** [comment]
Hello @jkschneider ,

I don't think so, I was quoting the full property name in my first post and as I said, I can see while debugging the MeterRegistry that that specific configuration value is correctly read.

The underlying problem is that - according to me - the DEFAULT HierarchicalNameMapper just doesn't use that prop, and that it is the default instance created in the GraphiteMetricsExportAutoConfiguration .

**Comment by jkschneider** [comment]
Ah, sorry @jceloi. Thanks for the detailed report.

**Comment by snicoll** [comment]
@jkschneider I've attempted to fix this bug in 8f4e27e and I have a question. The check on a custom `HierarchicalNameMapper` looks fragile to me. Is the intended purpose is that as soon as you define your own `HierarchicalNameMapper` it has to be able to support all monitoring systems that you want to use? If that's the contract, then I guess my fix should work, a review would be nice regardless.

**Comment by snicoll** [comment]
If we decide to do #12676, we can first merge this one and then remove the exposure of `HierarchicalNameMapper` altogether.
